### DISCUSSION TITLE
Use Dev Container CLI for building pre-built base image

### DISCUSSION THREAD

**Discussion by eitsupi** [PRQL/prql#2898] [General]
`docker/build-push-action` is currently used to build the image. I remember this was to use gha cache.
https://github.com/PRQL/prql/blob/d5d5362722985593db50bb67c00c7594217702be/.github/workflows/build-devcontainer.yaml#L62-L64

However, I suspect that the GHA cache is not really important here, since in practice this repository always exceeds the 10 GB limit of the GHA cache and the old cache will soon be unusable.
So I am wondering if we can switch to building with the Dev Container CLI, which can build images from devcontainer.json instead of not being able to use the GHA cache.
This would allow us to install note.js and Python, which are currently installed on the user side.

**Comment by max-sixty** [comment]
We can definitely try!

Does the devcontainer CLI build a docker image? I agree it would be good to have python & node built on the server rather than the client.

While the GHA cache indeed isn't helpful with our current setup, the docker repository should be reasonably helpful for caching. It should be caching the early layers there, so we avoid rebuilding them if they haven't changed. But:
- given the taskfike is opaque, we end up rebuilding anything coming from that on any change...
- I'm not sure whether it's working well — I think when I looked a while ago it might not have been caching any early layers. It should be picking up the `cache_from` reference to our image, in addition to the GHA cache.

...so maybe we shouldn't put much value on the cache, and using the devcontainer CLI has fewer downsides.

**Reply by eitsupi** [reply]
Ok, thanks. I will look at this later.

> Does the devcontainer CLI build a docker image?

Dev Container CLI calls the `docker buildx build` command, so it is buildx that actually does the build.
It just does not have access to features like GHA cache because the Dev Container CLI does not currently support all the options of the `buildx build` command.

**Reply by max-sixty** [reply]
OK great. I had a look at https://github.com/devcontainers/cli, it looks like it does indeed build an image. So this does seem promising. It should also be able to use the docker cache, so there's no reason not to try this!

(I'm trying to remember why we didn't do this last time — maybe we placed a lot of weight on the GitHub cache in our decision?)

**Reply by eitsupi** [reply]
> (I'm trying to remember why we didn't do this last time — maybe we placed a lot of weight on the GitHub cache in our decision?)

I think so (in #1893)

**Reply by max-sixty** [reply]
Hmmm, looks like this was the 1st of two options in https://github.com/PRQL/prql/pull/1893#issuecomment-1444449566, and we thought we could start with the current setup and grow into the integrated approach. So we're on track :)
